parser-typechecker/src/Unison/Codebase/FileCodebase/Common.hs

```diff
@@ -0,0 +1,601 @@
+import UnliftIO.IO.File (writeBinaryFile)
+
+type CodebasePath = FilePath
+
+data Err
+  = InvalidBranchFile FilePath String
+  | InvalidEditsFile FilePath String
+  | NoBranchHead FilePath
+  | CantParseBranchHead FilePath
+  | AmbiguouslyTypeAndTerm Reference.Id
+  | UnknownTypeOrTerm Reference
+  deriving Show
+
+type SimpleLens s a = Lens s s a a
+
+codebasePath :: FilePath
+codebasePath = ".unison" </> "v1"
+
+formatAnn :: S.Format Ann
+formatAnn = S.Format (pure External) (\_ -> pure ())
+
+type SyncToDir m v a
+  = S.Format v
+  -> S.Format a
+  -> CodebasePath
+  -> CodebasePath
+  -> Branch m
+  -> m (Branch m)
+
+termsDir, typesDir, branchesDir, branchHeadDir, editsDir
+  :: CodebasePath -> FilePath
+termsDir root = root </> codebasePath </> "terms"
+typesDir root = root </> codebasePath </> "types"
+branchesDir root = root </> codebasePath </> "paths"
+branchHeadDir root = branchesDir root </> "_head"
+editsDir root = root </> codebasePath </> "patches"
+
+termDir, declDir :: CodebasePath -> Reference.Id -> FilePath
+termDir root r = termsDir root </> componentIdToString r
+declDir root r = typesDir root </> componentIdToString r
+
+referenceToDir :: Reference -> FilePath
+referenceToDir r = case r of
+  Reference.Builtin name -> "_builtin" </> encodeFileName (Text.unpack name)
+  Reference.DerivedId hash -> componentIdToString hash
+
+dependentsDir', typeIndexDir', typeMentionsIndexDir' :: FilePath -> FilePath
+
+dependentsDir :: CodebasePath -> Reference -> FilePath
+dependentsDir root r = dependentsDir' root </> referenceToDir r
+dependentsDir' root = root </> codebasePath </> "dependents"
+
+watchesDir :: CodebasePath -> Text -> FilePath
+watchesDir root UF.RegularWatch =
+  root </> codebasePath </> "watches" </> "_cache"
+watchesDir root kind =
+  root </> codebasePath </> "watches" </> encodeFileName (Text.unpack kind)
+watchPath :: CodebasePath -> UF.WatchKind -> Reference.Id -> FilePath
+watchPath root kind id =
+  watchesDir root (Text.pack kind) </> componentIdToString id <> ".ub"
+
+typeIndexDir :: CodebasePath -> Reference -> FilePath
+typeIndexDir root r = typeIndexDir' root </> referenceToDir r
+typeIndexDir' root = root </> codebasePath </> "type-index"
+
+typeMentionsIndexDir :: CodebasePath -> Reference -> FilePath
+typeMentionsIndexDir root r = typeMentionsIndexDir' root </> referenceToDir r
+typeMentionsIndexDir' root = root </> codebasePath </> "type-mentions-index"
+
+decodeFileName :: FilePath -> String
+decodeFileName = go where
+  go ('$':tl) = case span (/= '$') tl of
+    ("forward-slash", _:tl) -> '/' : go tl
+    ("back-slash", _:tl) ->  '\\' : go tl
+    ("colon", _:tl) -> ':' : go tl
+    ("star", _:tl) -> '*' : go tl
+    ("question-mark", _:tl) -> '?' : go tl
+    ("double-quote", _:tl) -> '\"' : go tl
+    ("less-than", _:tl) -> '<' : go tl
+    ("greater-than", _:tl) -> '>' : go tl
+    ("pipe", _:tl) -> '|' : go tl
+    ('x':hex, _:tl) -> decodeHex hex ++ go tl
+    ("",_:tl) -> '$' : go tl
```

Comment:
Wait what? Don’t understand your comment, isn’t that currently a builtin name? Likewise Num.+ etc

Reply:
Err, right — by working I mean surviving encode-decode round trips.  Currently it's "working" because there are no round trips, only "encode"; but this PR introduced some, which allowed me to notice the bug:
```
✔ ~/unison6 [master|…28⚑ 1] 
11:20 $ stack repl
λ> let roundTrip = decodeFileName . encodeFileName
λ> roundTrip . roundTrip . roundTrip $ "Nat.+"
"Nat.+" 👍 
λ> roundTrip . roundTrip . roundTrip $ "Nat.<"
"Nat.$$$$less-than$$$$"  👎 
```

I see now that there are two calls to `decodeFileName` in master, but they are there by mistake, don't do anything, and can be removed. The existing unit tests for roundtrip tests stuff like `<` but not stuff like `Universal.<`.

Reply:
Actually a bunch of those tests are marked "pending".

Reply:
I'm gonna fix up the tests.